executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+use super::compiler::{written_vars_in_effect, WitgenFunctionParams};
+use super::effect::{Assertion, Effect};
+
+use super::symbolic_expression::{
+    BinaryOperator, BitOperator, RPNExpression, RPNExpressionElem, UnaryOperator,
+};
+use super::variable::{Cell, Variable};
+use crate::witgen::data_structures::mutable_state::MutableState;
+use crate::witgen::jit::effect::MachineCallArgument;
+use crate::witgen::machines::LookupCell;
+use crate::witgen::QueryCallback;
+use powdr_number::{FieldElement, LargeInt};
+
+use std::collections::HashMap;
+
+// Witgen effects compiled into interpreter instructions.
+pub struct EffectsInterpreter<T: FieldElement> {
+    first_column_id: u64,
+    column_count: usize,
+    vars: usize,
+    actions: Vec<InterpreterAction<T>>,
+}
+
+// Witgen effects compiled into "instructions".
+// Variables have been removed and replaced by their index in the variable list.
+enum InterpreterAction<T: FieldElement> {
+    ReadCell(usize, Cell),
+    ReadParam(usize, usize),
+    AssignExpression(usize, RPNExpression<T, usize>),
+    WriteCell(usize, Cell),
+    WriteParam(usize, usize),
+    WriteKnown(Cell),
+    MachineCall(Vec<usize>, u64, Vec<MachineCallArgument<T, usize>>),
+    Assertion(RPNExpression<T, usize>, RPNExpression<T, usize>, bool),
+}
+
+impl<T: FieldElement> EffectsInterpreter<T> {
+    pub fn new(
+        first_column_id: u64,
+        column_count: usize,
+        known_inputs: &[Variable],
+        effects: &[Effect<T, Variable>],
+    ) -> Self {
+        let mut actions = vec![];
+        let mut var_idx = HashMap::new();
+        let mut vars = 0;
+
+        // map variables to indices, so they can be quickly accessed as sequential memory (kept in a vec) during execution
+        let mut map_var_idx = |var: &Variable| -> usize {
+            let idx = *var_idx.entry(var.clone()).or_insert_with(|| {
+                let idx = vars;
+                vars += 1;
+                idx
+            });
+            idx
+        };
+
+        // load known inputs
```

Comment:
Sounds like this code would be perfect inside a function called `load_known_inputs` ;)